### PR TITLE
20414 - EFT Short name summaries search

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/20414

*Description of changes:*
- new route for eft short name summary search to support new UI table

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/resources/v1/eft_short_names.py [hunk 1]
```diff
@@ -73,6 +74,36 @@ def get_eft_shortnames():
...
+@bp.route('/summaries', methods=['GET', 'OPTIONS'])
+@cross_origin(origins='*', methods=['GET'])
+@_jwt.requires_auth
+@_jwt.has_one_of_roles([Role.SYSTEM.value, Role.MANAGE_EFT.value])
+def get_eft_shortname_summaries():
+    """Get all eft short name summaries."""
+    current_app.logger.info('<get_eft_shortname_summaries')
+    page: int = int(request.args.get('page', '1'))
+    limit: int = int(request.args.get('limit', '10'))
+    short_name = request.args.get('shortName', None)
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/eft_short_names.py:86]: perhaps candidate for a function + tuple return?
`page, limit, short_name = get_request_args(['page','limit','shortName'])`

File: pay-api/src/pay_api/resources/v1/eft_short_names.py [hunk 2]
```diff
@@ -73,6 +74,36 @@ def get_eft_shortnames():
...
+    credits_remaining = request.args.get('creditsRemaining', None)
+    linked_accounts_count = request.args.get('linkedAccountsCount', None)
+    payment_received_start_date = request.args.get('paymentReceivedStartDate', None)
+    payment_received_end_date = request.args.get('paymentReceivedEndDate', None)
+
+    response, status = EFTShortnameSummariesService.search(EFTShortnamesSearch(
+        id=short_name_id,
+        deposit_start_date=string_to_date(payment_received_start_date),
+        deposit_end_date=string_to_date(payment_received_end_date),
+        credit_remaining=Decimal(credits_remaining) if credits_remaining is not None else None,
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/eft_short_names.py:97]: helper function?

File: pay-api/src/pay_api/resources/v1/eft_short_names.py [hunk 3]
```diff
@@ -73,6 +74,36 @@ def get_eft_shortnames():
...
+        linked_accounts_count=int(linked_accounts_count) if linked_accounts_count is not None else None,
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/eft_short_names.py:98]: `int(linked_accounts_count or 0)` probably easier here

File: pay-api/src/pay_api/services/eft_short_name_summaries.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    @classmethod
+    def search(cls, search_criteria: EFTShortnamesSearch):
+        """Search eft short name summaries."""
+        current_app.logger.debug('<search')
+        search_count = cls.get_search_count()
+        search_query = cls.get_search_query(search_criteria)
+        pagination = search_query.paginate(per_page=search_criteria.limit,
+                                           page=search_criteria.page)
+
+        summary_list = [EFTSummarySchema.from_row(short_name_summary) for short_name_summary in pagination.items]
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_short_name_summaries.py:43]: I've seen this code repeated in a few services, might be up for helper function?

File: pay-api/src/pay_api/services/eft_short_name_summaries.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                       .group_by(EFTShortnameModel.id).with_entities(EFTShortnameModel.id))
+
+        current_app.logger.debug('>get_search_count')
+        return count_query.count()
+
+    @classmethod
+    def get_search_query(cls, search_criteria: EFTShortnamesSearch):
+        """Query for short names based on search criteria."""
+        linked_account_subquery = cls.get_linked_count_query().subquery()
+        credit_remaining_subquery = cls.get_remaining_credit_query().subquery()
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_short_name_summaries.py:103]: Is it worth moving this into the relationship so we get it for free?

Comment (seeker25) [reply pay-api/src/pay_api/services/eft_short_name_summaries.py:103]: Perhaps something like this? 

https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html

So we don't need to use the service to query the credits etc.

**Review by seeker25 — APPROVED** [review]
LGTM, a few suggestions

File: pay-api/src/pay_api/services/eft_short_name_summaries.py (hunk 5 above)
Comment (seeker25) [reply pay-api/src/pay_api/services/eft_short_name_summaries.py:103]: EG `relationship(Child, primaryjoin=Child.parent_id == Parent.id)` not sure if that's possible

File: pay-api/src/pay_api/services/eft_short_name_summaries.py (hunk 4 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/eft_short_name_summaries.py]: Added an unstructure helper.

I have some work on another EFT search route, I will do some clean up there and update anywhere that can use this helper.

File: pay-api/src/pay_api/resources/v1/eft_short_names.py (hunk 1 above)
Comment (ochiu) [reply pay-api/src/pay_api/resources/v1/eft_short_names.py:85]: Will ponder this one a bit more, it starts getting clunky when we consider defaults along with a longer list of params.

File: pay-api/src/pay_api/services/eft_short_name_summaries.py [hunk 6]
```diff
@@ -40,9 +40,7 @@ def search(cls, search_criteria: EFTShortnamesSearch):
         pagination = search_query.paginate(per_page=search_criteria.limit,
                                            page=search_criteria.page)
 
-        summary_list = [EFTSummarySchema.from_row(short_name_summary) for short_name_summary in pagination.items]
-        converter = Converter()
-        summary_list = converter.unstructure(summary_list)
+        summary_list = unstructure_schema_items(EFTSummarySchema, pagination.items)
```
Comment (seeker25) [inline pay-api/src/pay_api/services/eft_short_name_summaries.py:43]: so much nicer

**Pull request merged**